MoltenVK/MoltenVK/GPUObjects/MVKDevice.mm

```diff
@@ -2668,12 +2678,30 @@ static uint32_t mvkGetEntryProperty(io_registry_entry_t entry, CFStringRef prope
 
 MVKSemaphore* MVKDevice::createSemaphore(const VkSemaphoreCreateInfo* pCreateInfo,
 										 const VkAllocationCallbacks* pAllocator) {
-	if (_useMTLFenceForSemaphores) {
-		return new MVKSemaphoreMTLFence(this, pCreateInfo);
-	} else if (_useMTLEventForSemaphores) {
-		return new MVKSemaphoreMTLEvent(this, pCreateInfo);
+	const VkSemaphoreTypeCreateInfo* pTypeCreateInfo = nullptr;
+	for (auto* next = (const VkBaseInStructure*)pCreateInfo->pNext; next; next = next->pNext) {
+		switch (next->sType) {
+			case VK_STRUCTURE_TYPE_SEMAPHORE_TYPE_CREATE_INFO:
+				pTypeCreateInfo = (VkSemaphoreTypeCreateInfo*)next;
+				break;
+			default:
+				break;
+		}
+	}
+	if (pTypeCreateInfo && pTypeCreateInfo->semaphoreType == VK_SEMAPHORE_TYPE_TIMELINE) {
+		if (_pMetalFeatures->events) {
+			return new MVKTimelineSemaphoreMTLEvent(this, pCreateInfo, pTypeCreateInfo);
+		} else {
+			return new MVKTimelineSemaphoreEmulated(this, pCreateInfo, pTypeCreateInfo);
+		}
 	} else {
-		return new MVKSemaphoreEmulated(this, pCreateInfo);
+		if (_useMTLFenceForSemaphores) {
+			return new MVKSemaphoreMTLFence(this, pCreateInfo);
+		} else if (_useMTLEventForSemaphores) {
+			return new MVKSemaphoreMTLEvent(this, pCreateInfo);
+		} else {
+			return new MVKSemaphoreEmulated(this, pCreateInfo);
+		}
```

Comment:
Currently, by default, `_useMTLFenceForSemaphores` is enabled and `_useMTLEventForSemaphores` is disabled, which means regular `VkSemaphores` will prefer using `MTLFence` over `MTLEvent`.

Given that if a timeline semaphore is requested, we must use `MTLSharedEvent`, should we re-evaluate and possibly flip this? Is there any drive or appetite to prefer using `MTLEvent` over `MTLFence` for regular `VkSemaphores`?

Reply:
IIRC the reason for preferring `MTLFence` for binary semaphores is that we observed problems when `MTLEvent`s were used with presentation, where waits would time out and subsequently cause the system to revoke access to the device. The presentation code has changed significantly since then, but I'm still investigating whether or not it's safe to start using `MTLEvent` again.

(I keep forgetting that most people don't have memories as long as mine. How ironic.)

Reply:
Now where was that issue again...?

Oh yeah, there were two, #602 and #625.

Reply:
>I keep forgetting that most people don't have memories as long as mine. How ironic. (🤦🏻‍♂️...for the irony joke).

Yeah...raw memory is definitely not my strong suit. I'd not survive long in medical school. 😉

Sorry...I didn't expect you to troll back into the previous change triggers...but was more looking for your thoughts on whether either our code has evolved...or Apple's `MTLEvent` code has...based on the sync testing you were doing.

>The presentation code has changed significantly since then, but I'm still investigating whether or not it's safe to start using MTLEvent again.

Okay. Thanks for the review. If you do continue to investigate and it reveals that `MTLEvents` make sense, we can flip back.